### PR TITLE
Use HTTP Basic authentication for worker in MsgPack protocol

### PR DESCRIPTION
**Pull request by mokibit**
This PR introduces authentication with HTTP Basic "Authorization" header.

* [x] I have updated the unit tests
* [not needed] I have created a file in the `newsfragments` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: master/buildbot/worker/protocols/manager/msgpack.py [hunk 1]
```diff
@@ -344,10 +355,38 @@ def get_message_result(self, msg):
...
+    @defer.inlineCallbacks
     def onConnect(self, request):
         if self.debug:
             log.msg(f"Client connecting: {request.peer}")
 
+        try:
+            value = request.headers['authorization']
+            username, password = decode_http_authorization_header(value)
+        except Exception as e:
+            raise ConnectionDeny(401, "Unauthorized") from e
```
Comment (tardyp) [inline master/buildbot/worker/protocols/manager/msgpack.py:367]: if there is no autentication header, then we should throw 401 Unauthorized

If the authentication header is malformated: 400 Bad request

File: master/buildbot/worker/protocols/manager/msgpack.py [hunk 2]
```diff
@@ -344,10 +355,38 @@ def get_message_result(self, msg):
...
+                    authentication = False
+            else:
+                authentication = False
+        except Exception as e:
+            raise Exception("Internal error") from e
+        finally:
+            eventually(dispatcher.master.initLock.release)
+
+        if not authentication:
+            raise ConnectionDeny(401, "Unauthorized")
```
Comment (tardyp) [inline master/buildbot/worker/protocols/manager/msgpack.py:388]: There is no need for authentication flag. just raise 401 at the point of the double else

File: master/buildbot/worker/protocols/manager/msgpack.py [hunk 3]
```diff
@@ -344,10 +355,38 @@ def get_message_result(self, msg):
...
+                pwd, _ = dispatcher.users[username]
+                if pwd == password:
+                    self.worker_name = username
+                    authentication = True
+                else:
+                    authentication = False
+            else:
+                authentication = False
+        except Exception as e:
+            raise Exception("Internal error") from e
```
Comment (tardyp) [inline master/buildbot/worker/protocols/manager/msgpack.py:383]: catching bare exception do not add anything

File: master/buildbot/worker/protocols/manager/msgpack.py [hunk 4]
```diff
@@ -344,10 +355,38 @@ def get_message_result(self, msg):
...
+        finally:
```
Comment (tardyp) [inline master/buildbot/worker/protocols/manager/msgpack.py:384]: you can keep the finally without any except statement

File: master/buildbot/worker/protocols/manager/msgpack.py [hunk 5]
```diff
@@ -344,10 +355,38 @@ def get_message_result(self, msg):
...
+
+        try:
+            dispatcher = self.get_dispatcher()
+            yield dispatcher.master.initLock.acquire()
+
+            if username in dispatcher.users:
+                pwd, _ = dispatcher.users[username]
+                if pwd == password:
+                    self.worker_name = username
+                    authentication = True
```
Comment (tardyp) [inline master/buildbot/worker/protocols/manager/msgpack.py:377]: no need for a flag here. please simplify

**Review by tardyp — CHANGES_REQUESTED** [review]
there are some rooms for simplification in the authentication method.
Please make it as simple as possible to limit the attack surface

File: master/buildbot/worker/protocols/manager/msgpack.py (hunk 2 above)
Comment (mokibit) [reply master/buildbot/worker/protocols/manager/msgpack.py:388]: Authentication flag is needed to strictly separate authentication errors from other errors that can appear in the code and give too much information for potential adversary. See my comment below.

File: master/buildbot/worker/protocols/manager/msgpack.py (hunk 3 above)
Comment (mokibit) [reply master/buildbot/worker/protocols/manager/msgpack.py:383]: This exception catches any possible errors which are not related to authentication. Therefore, if an error appears, the adversary will not be able to see any additional information about the error so the attack surface is limited.

**Comment by p12tic** [comment]
Ping @tardyp, @mokibit raised some good points.

**Comment by p12tic** [comment]
@mokibit The title of the PR is not clear, seems like it's introducing this new authentication for pb protocol too.

File: master/buildbot/worker/protocols/manager/msgpack.py (hunk 1 above)
Comment (mokibit) [reply master/buildbot/worker/protocols/manager/msgpack.py]: Fixed.

**Comment by mokibit** [comment]
@p12tic 
> The title of the PR is not clear, seems like it's introducing this new authentication for pb protocol too.

Fixed

**Review by p12tic — APPROVED** [review]

**Comment by p12tic** [comment]
@mokibit Please rebase on top of latest master.

**Pull request merged**